rustc nightly (7480389 2022-08-25) gives erroneous fix advice on mismatching if-branch types

rustc version: rustc 1.65.0-nightly (748038961 2022-08-25)

I have a very small reproducer for a weird fix advice from rustc when the two branches of an if had a type mismatch:
```
fn main() {
    let x = if true {
        Ok(Ok(()))
    } else {
        Ok(())
    };
}
```
This will produce the following output:
```
error[E0308]: `if` and `else` have incompatible types
 --> weird-fix-advice-else-ok.rs:5:9
  |
2 |       let x = if true {
  |  _____________-
3 | |         Ok(Ok(()))
  | |         ---------- expected because of this
4 | |     } else {
5 | |         Ok(())
  | |         ^^^^^^ expected enum `Result`, found `()`
6 | |     };
  | |_____- `if` and `else` have incompatible types
  |
  = note: expected enum `Result<Result<(), _>, _>`
             found enum `Result<(), _>`
help: try wrapping the expression in a variant of `Result`
  |
4 ~     } else Ok({
5 |         Ok(())
6 ~     });
  |
4 ~     } else Err({
5 |         Ok(())
6 ~     });
  |

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
```
Applying this fix obviously doesn't help:
```
error: expected `{`, found `Ok`
 --> weird-fix-advice-else-ok.rs:4:12
  |
4 |     } else Ok({
  |            ^^ expected `{`
  |
help: try placing this code inside a block
  |
4 ~     } else { Ok({
5 |         Ok(())
6 ~     }) };
  |

error: aborting due to previous error
```

--- a/compiler/rustc_typeck/src/check/demand.rs
+++ b/compiler/rustc_typeck/src/check/demand.rs
@@ -417,6 +417,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     hir::def::CtorKind::Const => unreachable!(),
                 };
 
+                // Suggest constructor as deep into the block tree as possible.
+                // This fixes https://github.com/rust-lang/rust/issues/101065,
+                // and also just helps make the most minimal suggestions.
+                let mut expr = expr;
+                while let hir::ExprKind::Block(block, _) = &expr.kind
+                    && let Some(expr_) = &block.expr
+                {
+                    expr = expr_
+                }
+
                 vec![
                     (expr.span.shrink_to_lo(), format!("{prefix}{variant}{open}")),
                     (expr.span.shrink_to_hi(), close.to_owned()),

--- a/src/test/ui/suggestions/issue-101065.fixed
+++ b/src/test/ui/suggestions/issue-101065.fixed
@@ -1,0 +1,14 @@
+// check-fail
+// run-rustfix
+
+enum FakeResult<T> {
+    Ok(T)
+}
+
+fn main() {
+    let _x = if true {
+        FakeResult::Ok(FakeResult::Ok(()))
+    } else {
+        FakeResult::Ok(FakeResult::Ok(())) //~ERROR E0308
+    };
+}

--- a/src/test/ui/suggestions/issue-101065.rs
+++ b/src/test/ui/suggestions/issue-101065.rs
@@ -1,0 +1,14 @@
+// check-fail
+// run-rustfix
+
+enum FakeResult<T> {
+    Ok(T)
+}
+
+fn main() {
+    let _x = if true {
+        FakeResult::Ok(FakeResult::Ok(()))
+    } else {
+        FakeResult::Ok(()) //~ERROR E0308
+    };
+}

--- a/src/test/ui/suggestions/issue-101065.stderr
+++ b/src/test/ui/suggestions/issue-101065.stderr
@@ -1,0 +1,23 @@
+error[E0308]: `if` and `else` have incompatible types
+  --> $DIR/issue-101065.rs:12:9
+   |
+LL |       let _x = if true {
+   |  ______________-
+LL | |         FakeResult::Ok(FakeResult::Ok(()))
+   | |         ---------------------------------- expected because of this
+LL | |     } else {
+LL | |         FakeResult::Ok(())
+   | |         ^^^^^^^^^^^^^^^^^^ expected enum `FakeResult`, found `()`
+LL | |     };
+   | |_____- `if` and `else` have incompatible types
+   |
+   = note: expected enum `FakeResult<FakeResult<()>>`
+              found enum `FakeResult<()>`
+help: try wrapping the expression in `FakeResult::Ok`
+   |
+LL |         FakeResult::Ok(FakeResult::Ok(()))
+   |         +++++++++++++++                  +
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

